storage/repair_queue.go

```diff
@@ -0,0 +1,135 @@
+
+package storage
+
+import (
+	"time"
+
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/hlc"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+const (
+	// repairQueueMaxSize is the max size of the replicate queue.
+	repairQueueMaxSize = 100
+
+	// replicaQueueTimerDuration is the duration between removals of queued
+	// replicas.
+	repairQueueTimerDuration = 0 // zero duration to process removals greedily
+)
+
+// repairQueue manages a queue of replicas which may have one of its other
+// replica on a dead store.
+type repairQueue struct {
+	*baseQueue
+	storePool      *StorePool
+	replicateQueue *replicateQueue
+	clock          *hlc.Clock
+}
+
+// makeRepairQueue returns a new instance of repairQueue.
+func makeRepairQueue(storePool *StorePool, replicateQueue *replicateQueue, clock *hlc.Clock) repairQueue {
+	rq := repairQueue{
+		storePool:      storePool,
+		replicateQueue: replicateQueue,
+		clock:          clock,
+	}
+	// rq must be a pointer in order to setup the reference cycle.
+	rq.baseQueue = newBaseQueue("repair", &rq, repairQueueMaxSize)
+	return rq
+}
+
+// needsLeaderLease implements the queueImpl interface.
+// This should not require a leader lease, but if we find that there is too
+// much contention on the replica removals we can revisit that decision.
+func (rq repairQueue) needsLeaderLease() bool {
+	return false
+}
+
+// shouldQueue implements the queueImpl interface.
+func (rq repairQueue) shouldQueue(now proto.Timestamp, repl *Replica) (shouldQ bool, priority float64) {
+	shouldQ, priority, _ = rq.needsRepair(repl)
+	return
+}
+
+// needsRepair determines if any replicas that belong to the same range as repl
+// belong to a dead store. If so, it returns the priority (the number of
+// replicas on dead stores) and the list of these 'dead' replicas.
+func (rq repairQueue) needsRepair(repl *Replica) (bool, float64, []proto.Replica) {
+	rangeDesc := repl.Desc()
+	var deadReplicas []proto.Replica
+	for _, otherRepl := range rangeDesc.Replicas {
+		if rq.storePool.getStoreDetail(otherRepl.StoreID).dead {
+			deadReplicas = append(deadReplicas, otherRepl)
+		}
+	}
+
+	if len(deadReplicas) == 0 {
+		return false, float64(0), deadReplicas
+	}
+
+	// Does the range still have a quorum of live replicas? We shouldn't remove
+	// one without at least a quorum.
+	quorum := (len(rangeDesc.Replicas) / 2) + 1
+	liveReplicas := len(rangeDesc.Replicas) - len(deadReplicas)
+
+	// Priority is calculated based on need. By basing this off of the number
+	// of live replicas over he quorum amount ensures that in the case where a
+	// 5 replica range is only missing one it gets a lower priority than a 3
+	// replica range missing 1. Note that this may produce negative priorities,
+	// but these are correctly handled by baseQueue.
+	priority := float64(quorum - liveReplicas)
+	return liveReplicas >= quorum, priority, deadReplicas
```

Comment:
We won't be able to complete a remove replica operation without at least quorum.  Hence it's unavailable.